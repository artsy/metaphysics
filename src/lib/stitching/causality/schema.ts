import { createCausalityLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
  FilterRootFields,
  FilterTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

const blacklistedTypes: string[] = []
const whitelistedRootFields: string[] = []
const privateFields: string[] = ["lotStandingConnection"]
const permittedRootFields = [...whitelistedRootFields, ...privateFields]

export const executableCausalitySchema = () => {
  const causalityLink = createCausalityLink()
  const causalityTypeDefs = readFileSync("src/data/causality.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: causalityTypeDefs,
    link: causalityLink,
    buildSchemaOptions: {
      assumeValidSDL: true,
    },
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new FilterTypes((type) => !blacklistedTypes.includes(type.name)),
    new FilterRootFields((_operation, name, _field) => {
      return permittedRootFields.includes(name)
    }),
    new RenameTypes((name) => {
      if (name === "Long") return "Long"
      if (name === "Lot") name = "LotState"
      if (name === "Sale") name = "SaleState"
      return `Auctions${name}`
    }),
    new RenameRootFields((_operation, name) => {
      const capitalizedName = name.charAt(0).toUpperCase() + name.slice(1)
      if (privateFields.includes(name)) {
        return `_unused_auctions${capitalizedName}`
      }
      return `auctions${capitalizedName}`
    }),
  ])
}
