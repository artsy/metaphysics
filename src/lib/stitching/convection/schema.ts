import { createConvectionLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableConvectionSchema = async () => {
  const convectionLink = createConvectionLink()
  const convectionTypeDefs = readFileSync("src/data/convection.graphql", "utf8")

  // Setup the default Schema
  const schema = await makeRemoteExecutableSchema({
    schema: convectionTypeDefs,
    link: convectionLink,
  })

  // Remap the names of certain types from Convection to fit in the larger
  // metaphysics ecosystem.
  const remap = {
    Submission: "ConsignmentSubmission",
    Category: "SubmissionCategoryAggregation",
    Asset: "SubmissionCategoryAsset",
    State: "SubmissionStateAggregation",
  }

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      const newName = remap[name] || name
      return newName
    }),
  ])
}
