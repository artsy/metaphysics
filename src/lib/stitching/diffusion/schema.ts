import { createDiffusionLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableDiffusionSchema = () => {
  const diffusionLink = createDiffusionLink()
  const diffusionTypeDefs = readFileSync("src/data/diffusion.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: diffusionTypeDefs,
    link: diffusionLink,
  })

  // Remap the names of certain types from Diffusion to fit in the larger
  // metaphysics ecosystem.
  const remap = {}

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes((name) => {
      const newName = remap[name] || name
      return newName
    }),
  ])
}
