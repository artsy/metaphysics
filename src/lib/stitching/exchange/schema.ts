import { createExchangeLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"
import { ReplaceCommerceDateTimeType } from "./transformers/replaceCommerceDateTimeType"

export const executableExchangeSchema = (transforms) => {
  const exchangeSDL = readFileSync("src/data/exchange.graphql", "utf8")
  const exchangeLink = createExchangeLink()

  const schema = makeRemoteExecutableSchema({
    schema: exchangeSDL,
    link: exchangeLink,
  })

  // Return the new modified schema
  return transformSchema(schema, transforms)
  // Note that changes in this will need to be
}

export const transformsForExchange = [
  // Apply a prefix to all the typenames
  new RenameTypes((name) => {
    return `Commerce${name}`
  }),
  // Rename all the root fields to be camelCased
  new RenameRootFields(
    (_operation, name) =>
      `commerce${name.charAt(0).toUpperCase() + name.slice(1)}`
  ),
  // replace CommerceDateTime field with MP's dateField
  new ReplaceCommerceDateTimeType(),
]

export const legacyTransformsForExchange = [
  // Apply a prefix to all the typenames
  // for Legacy merged schema approach
  new RenameTypes((name) => {
    return `Ecommerce${name}`
  }),
  // Rename all the root fields to be camelCased
  new RenameRootFields(
    (_operation, name) =>
      `ecommerce${name.charAt(0).toUpperCase() + name.slice(1)}`
  ),
]
