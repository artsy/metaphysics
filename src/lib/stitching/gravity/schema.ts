import { createGravityLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  FilterTypes,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

const allowList = [
  "viewingRoom",
  "viewingRooms",
  "artistSeries",
  "artistSeriesConnection",
]

export const executableGravitySchema = () => {
  const gravityTypeDefs = readFileSync("src/data/gravity.graphql", "utf8")

  const gravityLink = createGravityLink()
  const schema = makeRemoteExecutableSchema({
    schema: gravityTypeDefs,
    link: gravityLink,
  })

  // Types which come from Gravity which MP already has copies of.
  // In the future, these could get merged into the MP types.
  const duplicatedTypes = [
    "Artist",
    "Artwork",
    "ArtistEdge",
    "ArtworkEdge",
    "ArtworkConnection",
    "ArtistConnection",
    "Money",
    "MoneyInput",
  ]

  // Types which come from Gravity that are not (yet) needed in MP.
  // In the future, these can be removed from this list as they are needed.
  const unusedTypes = [
    "DebitCommissionExemptionInput",
    "DebitCommissionExemptionPayload",
    "RefundCommissionExemptionInput",
    "RefundCommissionExemptionPayload",
  ]

  // Return the new modified schema
  return transformSchema(schema, [
    // Remove types which Metaphysics handles better
    new FilterTypes((type) => {
      return (
        !duplicatedTypes.includes(type.name) && !unusedTypes.includes(type.name)
      )
    }),
    // So, we cannot remove all of the types from a schema is a lesson I have
    // learned in creating these transformations. This means that there has to
    // be at least one type still inside the Schema (excluding the Mutation or
    // the Query)
    // When Partner was removed, we'd see this error
    // https://github.com/graphcool/graphql-import/issues/73
    // but I don't think we're exhibiting the same bug.
    new RenameTypes((name) => {
      if (name === "Partner") {
        return "DoNotUseThisPartner"
      }
      return name
    }),
    // We have the same restrictions for root, so let's prefix
    // for now
    new RenameRootFields((type, name, _field) => {
      if (type === "Query" && !allowList.includes(name)) {
        return `_unused_gravity_${name}`
      } else {
        return name
      }
    }),
  ])
}
