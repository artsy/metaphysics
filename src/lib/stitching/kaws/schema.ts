import { createKawsExecutor } from "./link"
import { wrapSchema, RenameTypes, RenameRootFields } from "@graphql-tools/wrap"
import { loadSchema } from "@graphql-tools/load"
import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader"

export const executableKawsSchema = async () => {
  const kawsExecutor = createKawsExecutor()

  // Setup the default Schema
  const schema = wrapSchema({
    schema: await loadSchema("src/data/kaws.graphql", {
      loaders: [new GraphQLFileLoader()],
    }),
    executor: kawsExecutor,
    transforms: [
      new RenameTypes((name) => {
        return `Marketing${name}`
      }),
      new RenameRootFields(
        (_operation, name) =>
          `marketing${name.charAt(0).toUpperCase() + name.slice(1)}`
      ),
    ],
  })

  return schema
}
