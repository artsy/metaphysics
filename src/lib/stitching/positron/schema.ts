import { createPositronLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executablePositronSchema = () => {
  const positronLink = createPositronLink()
  const positronTypeDefs = readFileSync("src/data/positron.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: positronTypeDefs,
    link: positronLink,
  })

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes((name) => {
      return `Editorial${name}`
    }),
    new RenameRootFields(
      (_operation, name) =>
        `editorial${name.charAt(0).toUpperCase() + name.slice(1)}`
    ),
  ])
}
