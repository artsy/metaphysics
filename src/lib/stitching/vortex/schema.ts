import { createVortexLink } from "./link"
import {
  makeRemoteExecutableSchema,
  transformSchema,
  RenameTypes,
  RenameRootFields,
  FilterRootFields,
} from "graphql-tools"
import { readFileSync } from "fs"

export const executableVortexSchema = ({
  removeRootFields = true,
}: { removeRootFields?: boolean } = {}) => {
  const vortexLink = createVortexLink()
  const vortexTypeDefs = readFileSync("src/data/vortex.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: vortexTypeDefs,
    link: vortexLink,
  })

  const rootFieldsToFilter = [
    "pricingContext",
    "partnerStat",
    "userStat",
    "BigInt",
  ]

  const filterTransform = new FilterRootFields(
    (_operation, name) => !rootFieldsToFilter.includes(name)
  )

  const transforms = [
    ...(removeRootFields ? [filterTransform] : []),
    new RenameTypes((name) => {
      if (
        name.includes("PriceInsight") ||
        name.includes("PageCursor") ||
        ["BigInt", "ISO8601DateTime"].includes(name)
      ) {
        return name
      } else {
        return `Analytics${name}`
      }
    }),
    new RenameRootFields((_operation, name) => {
      if (["priceInsights", "marketPriceInsights"].includes(name)) {
        return name
      } else {
        return `analytics${name.charAt(0).toUpperCase() + name.slice(1)}`
      }
    }),
  ]

  return transformSchema(schema, transforms)
}
