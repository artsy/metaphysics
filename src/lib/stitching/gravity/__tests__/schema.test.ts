import { executableGravitySchema } from "../schema"
import {
  getTypesFromSchema,
  getMutationFieldsFromSchema,
  getRootFieldsFromSchema,
} from "lib/stitching/lib/getTypesFromSchema"
import { printSchema } from "graphql"

it("Does not include gravity core types", async () => {
  const gravitySchema = await executableGravitySchema()
  const gravityTypes = await getTypesFromSchema(gravitySchema)

  expect(gravityTypes).not.toContain("Artist")
  expect(gravityTypes).not.toContain("Artwork")
  expect(gravityTypes).not.toContain("Partner")

  expect(gravityTypes).toContain("Mutation")
})

it("Includes the mutation recordArtworkView", async () => {
  const gravitySchema = await executableGravitySchema()
  const mutations = await getMutationFieldsFromSchema(gravitySchema)

  expect(mutations).toContain("recordArtworkView")
})

it("Does not include the root query fields", async () => {
  const gravitySchema = await executableGravitySchema()
  const rootFields = await getRootFieldsFromSchema(gravitySchema)

  expect(rootFields).not.toContain("artist")
  expect(rootFields).not.toContain("artists")
})

it("creates an SDL", async () => {
  const gravitySchema = await executableGravitySchema()
  expect(printSchema(gravitySchema, { commentDescriptions: true })).toBeTruthy()
})
