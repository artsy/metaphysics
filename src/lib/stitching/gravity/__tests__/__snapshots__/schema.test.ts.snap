// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates an SDL 1`] = `
"# App Authenticator Two-Factor Authentication factor
type AppSecondFactor implements SecondFactor {
  createdAt: ISO8601DateTime!
  disabledAt: ISO8601DateTime
  enabled: Boolean!
  enabledAt: ISO8601DateTime
  internalID: ID!
  kind: SecondFactorKind!
  name: String
  updatedAt: ISO8601DateTime!
}

# Backup Two-Factor Authentication factor
type BackupSecondFactor implements SecondFactor {
  createdAt: ISO8601DateTime!
  disabledAt: ISO8601DateTime
  enabled: Boolean!
  enabledAt: ISO8601DateTime
  internalID: ID!
  kind: SecondFactorKind!
  updatedAt: ISO8601DateTime!
}

# Autogenerated input type of CaptureHold
input CaptureHoldInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # Globally unique ID for the hold.
  inventoryHoldId: ID!

  # Globally unique ID for the requester of the hold. Ex. Exchange Order ID.
  referenceId: String!
}

# Autogenerated return type of CaptureHold
type CaptureHoldPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  holdOrError: InventoryHoldOrErrorUnion!
}

type ConditionReportRequest {
  internalID: ID!
  saleArtworkID: ID
  userID: ID
}

type DoNotUseThisPartner {
  displayName: String
  display_name: String @deprecated(reason: \\"Use displayName\\")
  givenName: String
  given_name: String @deprecated(reason: \\"Use givenName\\")
  id: ID!
  relativeSize: Int
  relative_size: Int @deprecated(reason: \\"Use relativeSize\\")
  slug: String
  subscriptionState: String
  subscription_state: String @deprecated(reason: \\"Use subscriptionState\\")
}

# A user-readable error
type Error {
  # Error code
  code: String!

  # Extra data about error.
  data: JSON

  # A description of the error
  message: String!

  # Which input value this error came from
  path: [String!]
}

# A user-readable list of errors
type Errors {
  # List of Errors.
  errors: [Error!]!
}

# Autogenerated input type of HoldInventory
input HoldInventoryInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  expiresInSeconds: Int
  items: [InventoryHoldItemAttributes!]!

  # Globally unique ID for the requester of this hold. Ex. Exchange Order ID.
  referenceId: String!
}

# Autogenerated return type of HoldInventory
type HoldInventoryPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  holdOrErrors: InventoryHoldOrErrorsUnion!
}

type InventoryHold {
  capturedAt: String
  id: ID!
  items(
    # Returns the elements in the list that come after the specified cursor.
    after: String

    # Returns the elements in the list that come before the specified cursor.
    before: String

    # Returns the first _n_ elements from the list.
    first: Int

    # Returns the last _n_ elements from the list.
    last: Int
  ): InventoryHoldItemConnection
  referenceId: String!
}

type InventoryHoldItem {
  artworkId: String!
  editionSetId: String
  id: ID!
  quantity: Int!
}

# Attributes for creating or updating an inventory hold item
input InventoryHoldItemAttributes {
  # ID of the artwork
  artworkId: String!

  # ID of edition set
  editionSetId: String
  quantity: Int!
}

# The connection type for InventoryHoldItem.
type InventoryHoldItemConnection {
  # A list of edges.
  edges: [InventoryHoldItemEdge]

  # A list of nodes.
  nodes: [InventoryHoldItem]

  # Information to aid in pagination.
  pageInfo: PageInfo!
}

# An edge in a connection.
type InventoryHoldItemEdge {
  # A cursor for use in pagination.
  cursor: String!

  # The item at the end of the edge.
  node: InventoryHoldItem
}

# A hold or error object
union InventoryHoldOrErrorsUnion = Errors | InventoryHold

# A hold or error object
union InventoryHoldOrErrorUnion = Error | InventoryHold

# An ISO 8601-encoded datetime
scalar ISO8601DateTime

# Represents untyped JSON
scalar JSON

# A money object with amount in minor currency and a currency code, or an error object
union MoneyOrErrorUnion = Error

type Mutation {
  captureHold(input: CaptureHoldInput!): CaptureHoldPayload
  holdInventory(input: HoldInventoryInput!): HoldInventoryPayload
  recordArtworkView(input: RecordArtworkViewInput!): RecordArtworkViewPayload
  requestConditionReport(input: RequestConditionReportInput!): RequestConditionReportPayload
}

# Information about pagination in a connection.
type PageInfo {
  # When paginating forwards, the cursor to continue.
  endCursor: String

  # When paginating forwards, are there more items?
  hasNextPage: Boolean!

  # When paginating backwards, are there more items?
  hasPreviousPage: Boolean!

  # When paginating backwards, the cursor to continue.
  startCursor: String
}

type Query {
  # Autocomplete resolvers.
  _unused_gravity_matchPartners(matchType: String, page: Int = 1, size: Int = 5, term: String!): [DoNotUseThisPartner!]

  # Autocomplete resolvers.
  _unused_gravity_match_partners(match_type: String, page: Int = 1, size: Int = 5, term: String!): [DoNotUseThisPartner!] @deprecated(reason: \\"Use matchPartners\\")

  # Find partners by IDs
  _unused_gravity_partners(ids: [ID!]!): [DoNotUseThisPartner!]

  # List enabled Two-Factor Authentication factors
  _unused_gravity_secondFactors(kinds: [SecondFactorKind!] = [app, sms, backup]): [SecondFactor!]!

  # Find a viewing room by ID
  viewingRoom(id: ID!): ViewingRoom
}

# Autogenerated input type of RecordArtworkView
input RecordArtworkViewInput {
  artwork_id: String!

  # A unique identifier for the client performing the mutation.
  clientMutationId: String
}

# Autogenerated return type of RecordArtworkView
type RecordArtworkViewPayload {
  # Id of viewed artwork
  artworkId: String!

  # Id of viewed artwork
  artwork_id: String! @deprecated(reason: \\"Use artworkId\\")

  # A unique identifier for the client performing the mutation.
  clientMutationId: String
}

# Autogenerated input type of RequestConditionReport
input RequestConditionReportInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # ID of the sale artwork.
  saleArtworkID: String!
}

# Autogenerated return type of RequestConditionReport
type RequestConditionReportPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  conditionReportRequest: ConditionReportRequest!
}

interface SecondFactor {
  createdAt: ISO8601DateTime!
  disabledAt: ISO8601DateTime
  enabled: Boolean!
  enabledAt: ISO8601DateTime
  internalID: ID!
  kind: SecondFactorKind!
  updatedAt: ISO8601DateTime!
}

# Two-Factor Authentication (2FA) Method
enum SecondFactorKind {
  # App authenticator 2FA method
  app

  # Backup 2FA method
  backup

  # SMS 2FA method
  sms
}

# SMS Two-Factor Authentication factor
type SmsSecondFactor implements SecondFactor {
  createdAt: ISO8601DateTime!
  disabledAt: ISO8601DateTime
  enabled: Boolean!
  enabledAt: ISO8601DateTime
  internalID: ID!
  kind: SecondFactorKind!
  lastDeliveredAt: ISO8601DateTime
  maskedPhone: String
  updatedAt: ISO8601DateTime!
}

# An artwork viewing room
type ViewingRoom {
  artworksConnection(
    # Returns the elements in the list that come after the specified cursor.
    after: String

    # Returns the elements in the list that come before the specified cursor.
    before: String

    # Returns the first _n_ elements from the list.
    first: Int

    # Returns the last _n_ elements from the list.
    last: Int
  ): ViewingRoomArtworkConnection

  # Body copy
  body: String

  # Datetime after which the viewing room is no longer viewable
  endAt: ISO8601DateTime

  # Image URL for hero image on viewing room home screen
  heroImageURL: String

  # Unique ID for this room
  internalID: ID!

  # Introductory paragraph
  introStatement: String

  # ID of the partner associated with this viewing room
  partnerID: String!
  pullQuote: String

  # Datetime when the viewing room is viewable
  startAt: ISO8601DateTime
  subsections: [ViewingRoomSubsection!]
  timeZone: String

  # Viewing room name
  title: String!
}

# An artwork associated with a viewing room. Note that the full artwork is stitched in in Metaphysics!
type ViewingRoomArtwork {
  artworkID: String!

  # Unique ID
  internalID: ID!
}

# The connection type for ViewingRoomArtwork.
type ViewingRoomArtworkConnection {
  # A list of edges.
  edges: [ViewingRoomArtworkEdge]

  # A list of nodes.
  nodes: [ViewingRoomArtwork]

  # Information to aid in pagination.
  pageInfo: PageInfo!
}

# An edge in a connection.
type ViewingRoomArtworkEdge {
  # A cursor for use in pagination.
  cursor: String!

  # The item at the end of the edge.
  node: ViewingRoomArtwork
}

# Title, image, text, and caption for a viewing room section
type ViewingRoomSubsection {
  # Body copy
  body: String

  # Image caption
  caption: String

  # Image URL
  imageURL: String

  # Unique ID for this subsection
  internalID: ID!

  # Section header
  title: String
}
"
`;
